### PR TITLE
Dynamic loading app update

### PR DESCRIPTION
**Pull request by anjiahao1**
## Summary
This PR replaces the dynamically loaded module_initialize function with the __attribute__((destructor)) keyword, and optimizes the compilation method for dynamic modules. Now, by simply adding the DYNLIB flag in the Makefile, the apps can be compiled as dynamic modules.
## Impact
sotest/modprint
## Testing
qemu with arm an547

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



**No, this PR does not fully meet the NuttX requirements.**

**Missing Information:**

* **Summary:**
    *  A clear explanation of *why* this change is necessary is missing. What specific problems does it solve? Is it a performance improvement, a bug fix, or a new feature? 
    *  What is the "functional part of the code" being changed? Be specific – mention the files/modules affected. 
* **Impact:**
    *  Many fields are incomplete. Address *all* impact categories. For example:
        *  **Is new feature added?** (If so, describe the feature) 
        *  **Impact on build:**  You mention changes in the Makefile. This requires a "YES" and a description of the changes. 
        *  **Impact on documentation:** Does this change necessitate documentation updates?
* **Testing:**
    *  **Insufficient detail:** 
       *  Provide specific OS and version information for your build host(s) (e.g., "Linux Ubuntu 20.04").
       *  "qemu with arm an547" is too vague.  Specify the QEMU version and the exact ARM architecture (e.g., ARMv7-M, ARMv8-A).
       *  **No testing logs:** You must include *actual* logs from your testing, both before and after the change, to demonstrate the issue and the fix. 

**Recommendations:**

1. **Expand the Summary:**  Provide a clear and concise rationale for the change.
2. **Complete the Impact Section:** Address all impact categories with specific details.
3. **Provide Detailed Testing Information:** Include OS versions, QEMU details, and actual testing logs. 

By addressing these points, your PR will be more informative and easier for reviewers to evaluate.

**Comment by anjiahao1** [comment]
this pr depends on [13819](https://github.com/apache/nuttx/pull/13819)

File: examples/module/chardev/chardev.c [hunk 1]
```diff
@@ -101,12 +101,13 @@ static ssize_t chardev_write(FAR struct file *filep, FAR const char *buffer,
  * Name: module_uninitialize
  ****************************************************************************/
 
-static int module_uninitialize(FAR void *arg)
+__attribute__((destructor))
```
Comment (patacongo) [inline examples/module/chardev/chardev.c:104]: > Isn't this specific to GCC compiler.  One of the basic rules is that we never include compiler-specific directories in common code that may use a different toolchain.

The correct way to handle this is to put a defininition in include/nuttx/compiler.h that defines a toolchain agnostic symbol for this.  For toolchains that do not support that attribute, that symbol can be defined as nothing.

Refer to "All Users Matter" in the to-level INVIOLABLES.txt

File: examples/sotest/sotest/sotest.c [hunk 2]
```diff
@@ -132,10 +111,10 @@ static void testfunc3(FAR const char *msg)
  * Name: module_uninitialize
  ****************************************************************************/
 
-static int module_uninitialize(FAR void *arg)
+__attribute__((destructor))
```
Comment (patacongo) [inline examples/sotest/sotest/sotest.c:114]: See comments elsewhere regarding the same use of this attribute

File: examples/sotest/modprint/modprint.c [hunk 3]
```diff
@@ -57,7 +40,8 @@ static const struct symtab_s g_modprint_exports[1] =
  * Name: modprint
  ****************************************************************************/
 
-static void modprint(FAR const char *fmt, ...)
+__attribute__((visibility("default")))
```
Comment (patacongo) [inline examples/sotest/modprint/modprint.c:43]: See comments for __attribute((destructor)).  This applies to the many occurrences of the visibility attribute.

File: examples/sotest/sotest/sotest.c [hunk 4]
```diff
@@ -150,14 +129,8 @@ static int module_uninitialize(FAR void *arg)
  *
  ****************************************************************************/
 
-int module_initialize(FAR struct mod_info_s *modinfo)
+__attribute__((constructor))
```
Comment (patacongo) [inline examples/sotest/sotest/sotest.c:132]: See comments in regard to __attribute__((destructor))

File: examples/module/chardev/chardev.c [hunk 5]
```diff
@@ -121,14 +122,9 @@ static int module_uninitialize(FAR void *arg)
  *
  ****************************************************************************/
 
-int module_initialize(FAR struct mod_info_s *modinfo)
+__attribute__((constructor))
```
Comment (patacongo) [inline examples/module/chardev/chardev.c:125]: See other comments with regard to the explicit use of GCC attributes in common code.

**Comment by anjiahao1** [comment]
depends on  [13852](https://github.com/apache/nuttx/pull/13852 )

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by anjiahao1** [comment]
look like ci build break 
image

**Comment by lupyuen** [comment]
I restarted the job, let's see if it works.

**Comment by xiaoxiang781216** [comment]
@anjiahao1 please fix:
```
====================================================================================
Configuration/Tool: canmv230/remote
2024-10-12 10:36:39
------------------------------------------------------------------------------------
  Cleaning...
  Configuring...
  Building NuttX...
/github/workspace/sources/apps /github/workspace/sources/nuttx
riscv-none-elf-ld: cannot open map file /github/workspace/sources/apps/binmap/init.map: No such file or directory
make[3]: *** [/github/workspace/sources/apps/Application.mk:304: /github/workspace/sources/apps/bin/sh] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [Makefile:53: /github/workspace/sources/apps/system/nsh_install] Error 2
make[1]: *** [Makefile:99: .import] Error 2
make: *** [Makefile:106: import] Error 2
/github/workspace/sources/nuttx
```

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**